src/Microsoft.ML.Recommend/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,403 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Internallearn;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Runtime.Recommend;
+using Microsoft.ML.Runtime.Recommend.Internal;
+
+[assembly: LoadableClass(typeof(MatrixFactorizationPredictor), null, typeof(SignatureLoadModel),
+    "Matrix Factorization Predictor Executor", MatrixFactorizationPredictor.LoaderSignature)]
+
+namespace Microsoft.ML.Runtime.Recommend
+{
+    public sealed class MatrixFactorizationPredictor : IPredictor, ICanSaveModel, ICanSaveInTextFormat, IMIValueMapper, ISchemaBindableMapper,
+        IUserHistoryToItemsRecommender
+    {
+        public const string LoaderSignature = "MFPredictorExec";
+        public const string RegistrationName = "MatrixFactorizationPredictor";
+
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                "MFPREDIC",
+                0x00010001, // Initial
+                0x00010001,
+                0x00010001,
+                "MFPredictorExec",
+                null);
+        }
+
+        private readonly IHost _host;
+        // The number of rows.
+        private readonly int _m;
+        // The number of columns.
+        private readonly int _n;
+        // The internal dimension.
+        private readonly int _k;
+        // The libMF is always single precision, so we will
+        // keep it as such even in the double build, even when
+        // the TLC output type is noted as being double.
+
+        // Packed _m by _k matrix.
+        private readonly Single[] _p;
+        // Packed _k by _n matrix.
+        private readonly Single[] _q;
+
+        private readonly KeyType _inputYType;
+        private readonly KeyType _inputXType;
+
+        public PredictionKind PredictionKind
+        {
+            get { return PredictionKind.Recommendation; }
+        }
+
+        public ColumnType OutputType { get { return NumberType.Float; } }
+        // REVIEW tfinley: Worth caching the below?
+        public ColumnType InputXType { get { return _inputXType; } }
+        public ColumnType InputYType { get { return _inputYType; } }
+
+        internal MatrixFactorizationPredictor(IHostEnvironment env, SafeTrainingAndModelBuffer buffer, KeyType xType, KeyType yType)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(RegistrationName);
+            _host.CheckValue(buffer, nameof(buffer));
+            _host.CheckValue(xType, nameof(xType));
+            _host.CheckValue(yType, nameof(xType));
+
+            buffer.Get(out _m, out _n, out _k, out _p, out _q);
+            _host.Assert(xType.RawKind == DataKind.U4);
+            _host.Assert(yType.RawKind == DataKind.U4);
+            _host.Assert(_n == xType.Count);
+            _host.Assert(_m == yType.Count);
```

Comment:
any sanity checks for k,p, and q?

Reply:
The following checks will be added:
            _host.Assert(_p.Length == _m * _k);
            _host.Assert(_q.Length == _n * _k);

---
In reply to: [225353449](https://github.com/dotnet/machinelearning/pull/1263#discussion_r225353449) [](ancestors = 225353449)